Content.Shared/_Crescent/Hardpoints/GridHardpointLimiterSystem.cs

```diff
@@ -0,0 +1,221 @@
+            // This is a bit hacky but necessary since BeforeAnchoredEvent isn't cancellable
+            Timer.Spawn(1, () => {
+                if (Exists(uid) && Transform(uid).Anchored)
+                {
+ //                   Logger.Warning($"[HardpointLimiter] Force-unanchoring hardpoint {uid} due to limit exceeded");
+                    _xform.Unanchor(uid);
+                }
+            });
+            return;
+        }
+
+//        Logger.Info($"[HardpointLimiter] FINAL CHECK PASSED: Allowing hardpoint {uid} to complete anchoring to grid {gridUid}");
+    }
+
+    private void OnHardpointUnanchorAttempt(EntityUid uid, HardpointComponent component, UnanchorAttemptEvent args)
+    {
+ //       Logger.Info($"[HardpointLimiter] OnHardpointUnanchorAttempt called for hardpoint {uid} - always allowing unanchor");
+        // Always allow unanchoring - we want to let people remove hardpoints
+    }
+
+    private void InitializeGridLimits(EntityUid grid)
+    {
+        if (!Exists(grid))
+            return;
+
+        // Create a simple server-side component for tracking
+        if (!TryComp<ServerGridHardpointTrackerComponent>(grid, out var limiter))
+        {
+            limiter = EntityManager.AddComponent<ServerGridHardpointTrackerComponent>(grid);
+        }
+
+
+        var found = 0;
+        var query = EntityQueryEnumerator<HardpointComponent, TransformComponent>();
+
+        while (query.MoveNext(out var uid, out _, out var xform))
+        {
+            // Only count anchored hardpoints on this specific grid
+            if (xform.GridUid == grid && xform.Anchored)
+                found++;
+        }
+
+        limiter.MaxHardpoints = found <= 1 ? 4 : found + 2;
+        limiter.CurrentHardpoints = found;
+
+        var gridName = TryComp<MetaDataComponent>(grid, out var meta) ? meta.EntityName : "Unknown";
+//        Logger.Info($"[HardpointLimiter] Grid {grid} ({gridName}) initialized with {found} hardpoints (max = {limiter.MaxHardpoints})");
+    }
+
+    private void OnCannonAnchored(HardpointCannonAnchoredEvent ev)
+    {
+ //       Logger.Info($"[HardpointLimiter] OnCannonAnchored called for cannon {ev.cannonUid} on grid {ev.gridUid}");
+
+        if (!TryComp<ServerGridHardpointTrackerComponent>(ev.gridUid, out var comp))
+        {
+ //           Logger.Warning($"[HardpointLimiter] No tracker component found for grid {ev.gridUid}!");
+            return;
+        }
+
+        // Recalculate current count to be sure
+        var actualCount = GetCurrentHardpointCount(ev.gridUid);
+        comp.CurrentHardpoints = actualCount;
+
+//        Logger.Info($"[HardpointLimiter] Grid {ev.gridUid} - Current: {comp.CurrentHardpoints}, Max: {comp.MaxHardpoints}");
+
+        if (comp.CurrentHardpoints > comp.MaxHardpoints)
+        {
+ //           Logger.Warning($"[HardpointLimiter] Hardpoint limit exceeded on grid {ev.gridUid}! Current: {comp.CurrentHardpoints}, Max: {comp.MaxHardpoints}");
+        }
+    }
```

Comment:
useless,  you care only about hardpoint not cannons taht get attached to them , this does nothing.